### PR TITLE
BUG: Make cobyla threadsafe

### LINKED ISSUE #9658: `scipy.optimize.minimize(method='COBYLA')` not threadsafe
I tried to run multiple optimisations with the cobyla method at the same time. Due to the asyncio based  communication with the modelling/calculation framework, I decided to use a multithreaded approach.

Platform/Distribution
* Python3.7.2/scipy-1.1.0  via conda on Debian testing
* Python3.7.2/scipy-1.1.0 via apt on Debian testing

In order to showcase my problems, I slimmed down the code to this example:
```
import concurrent.futures
import time
import scipy.optimize

def objective1(x):
    print(x, "1")
    time.sleep(0.1)
    return x[0]**2

def objective2(x):
    print(x, "2")
    time.sleep(0.1)
    return (x[0]-1)**2

min_method = "COBYLA"

def minimizer1():
    return scipy.optimize.minimize(objective1,
                                   [0.0],
                                   method=min_method)

def minimizer2():
    return scipy.optimize.minimize(objective2,
                                   [0.0],
                                   method=min_method)

with concurrent.futures.ThreadPoolExecutor() as pool:
    tasks = []
    tasks.append(pool.submit(minimizer1))
    tasks.append(pool.submit(minimizer2))

    for t in tasks:
        print(t.result())
```
(Yes, the original code used `args`, but I wanted to simplify it even more).

This code's output reads:
```
[0.] 1
[0.] 2
[1.] 2
[1.] 2
[0.125] 2
[2.] 2
[-0.25] 2
[1.5] 2
[0.125] 2
[0.75] 2
[-0.0625] 2
[1.125] 2
[0.03125] 2
[0.9375] 2
[-0.015625] 2
[1.03125] 2
[0.0078125] 2
[0.984375] 2
[-0.00390625] 2
[1.0078125] 2
[0.00195312] 2
[0.99609375] 2
[-0.00097656] 2
[1.00195312] 2
[0.00048828] 2
[0.99902344] 2
[-0.00024414] 2
[1.00048828] 2
[0.0001] 2
[0.99975586] 2
[-0.0001] 2
[1.0001] 2
     fun: 1.00020001
   maxcv: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 16
  status: 1
 success: True
       x: array([-0.0001])
Call-back cb_calcfc_in__cobyla__user__routines failed.
Segmentation fault
```

(Obviously) The segmentation fault is unexpected, but also looking at the sequence of values - features 32 lines and roughly alternating objective values from objective 1 and 2, but only objective 2 seems to be evaluated.

The stack trace is:
```
#0  0x00007f5e94d24dc0 in f2py_rout.cobyla_minimize () from /home/XXXX/lib/python3.7/site-packages/scipy/optimize/_cobyla.cpython-37m-x86_64-linux-gnu.so
#1  0x000055b6ce95146b in _PyObject_FastCallKeywords (callable=0x7f5e94d73670, stack=0x7f5e88001680, nargs=1, kwnames=0x7f5e94d74a08) at /tmp/build/80754af9/python_1546061345851/work/Objects/call.c:199
#2  0x000055b6ce9ac378 in call_function (kwnames=0x7f5e94d74a08, oparg=<optimized out>, pp_stack=<synthetic pointer>) at /tmp/build/80754af9/python_1546061345851/work/Python/ceval.c:4619
[....]
```

There are two ways to avoid the problems above:
* use a ProcessPoolExecutor
* use the minimizer SLSQP

Here the output for the ProcessPoolExecutor:
```
[0.] 1
[0.] 2
[1.] 1
[1.] 2
[2.] 2
[-1.] 1
[-0.5] 1
[1.5] 2
[0.25] 1
[0.75] 2
[1.125] 2
[-0.125] 1
[0.9375] 2
[0.0625] 1
[-0.03125] 1
[1.03125] 2
[0.015625] 1
[0.984375] 2
[1.0078125] 2
[-0.0078125] 1
[0.99609375] 2
[0.00390625] 1
[-0.00195312] 1
[1.00195312] 2
[0.00097656] 1
[0.99902344] 2
[-0.00048828] 1
[1.00048828] 2
[0.00024414] 1
[0.99975586] 2
[1.0001] 2
[-1.e-04] 1
[0.0001] 1
[0.9999] 2
     fun: 1e-08
   maxcv: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 17
  status: 1
 success: True
       x: array([0.0001])
     fun: 9.999999999997797e-09
   maxcv: 0.0
 message: 'Optimization terminated successfully.'
    nfev: 17
  status: 1
 success: True
       x: array([0.9999])
```
This output looks like I expect it.

I had a quick look at the Fortran/Python interface - sadly, I don't know enough about f2py to pin point the problem in the repository/code.

### PR DESCRIPTION
**Pull request by sturlamolden**
#### Reference issue

Closes gh-9658.

#### What does this implement/fix?

Cobyla is not threadsafe and can segfault when called from more than one thread. The problem is likely upstream in the f2py wrapper. Until the problem is fixed upstream in NumPy (f2py) we must synchronize the calls to cobyla with a lock.

#### Additional information

Related issue upstream:
https://github.com/numpy/numpy/issues/16513

### CODE REVIEW

File: scipy/optimize/cobyla.py [hunk 1]
```diff
@@ -22,9 +22,23 @@
...
-__all__ = ['fmin_cobyla']
+# Workaround as _cobyla.minimize is not threadsafe
+# due to an unknown f2py bug and can segfault, 
+# see gh-9658.
+
+from threading import Lock
+_module_lock = Lock()
+
+def synchronized(func):
+    def wrapper(*args, **kwargs):
```
Comment (pv) [inline scipy/optimize/cobyla.py:33]: @functools.wraps(func) here

File: scipy/optimize/cobyla.py [hunk 2]
```diff
@@ -22,9 +22,23 @@
     izip = zip
 
 
-__all__ = ['fmin_cobyla']
+# Workaround as _cobyla.minimize is not threadsafe
+# due to an unknown f2py bug and can segfault, 
+# see gh-9658.
+
+from threading import Lock
+_module_lock = Lock()
```
Comment (pv) [inline scipy/optimize/cobyla.py:30]: RLock, f2py is reentrant

File: scipy/optimize/cobyla.py [hunk 3]
```diff
@@ -22,9 +22,23 @@
     izip = zip
 
 
-__all__ = ['fmin_cobyla']
+# Workaround as _cobyla.minimize is not threadsafe
+# due to an unknown f2py bug and can segfault, 
+# see gh-9658.
+
+from threading import Lock
```
Comment (pv) [inline scipy/optimize/cobyla.py:29]: Import to top

File: scipy/optimize/cobyla.py [hunk 4]
```diff
@@ -22,9 +22,23 @@
...
+_module_lock = Lock()
+
+def synchronized(func):
+    def wrapper(*args, **kwargs):
+        with _module_lock:
+            return func(*args, **kwargs)
+    return wrapper
 
 
+__all__ = ['fmin_cobyla']
```
Comment (pv) [inline scipy/optimize/cobyla.py:39]: __all__ preferably just after imports

**Review by pv — CHANGES_REQUESTED** [review]

**Review by pv — CHANGES_REQUESTED** [review]
Test also should be added, probably the one from the issue works.

File: scipy/optimize/cobyla.py (hunk 1 above)
Comment (sturlamolden) [reply scipy/optimize/cobyla.py:38]: ok

File: scipy/optimize/cobyla.py (hunk 2 above)
Comment (sturlamolden) [reply scipy/optimize/cobyla.py]: It should be, but it is hard to say if it is the Fortran or C code that segfaults. Anyway, we are not doing any recursive calls here so both will work. I have changed it to RLock but I do not think it matters.

**Comment by sturlamolden** [comment]
> Test also should be added, probably the one from the issue works.

Included as requested.

**Comment by sturlamolden** [comment]
The CI failure is unrelated to this PR. Some problem with AR pack on ARM64.

**Comment by sturlamolden** [comment]
With upstream fix in https://github.com/numpy/numpy/pull/16519 the question is if we still need this, or at least it should be turned off for NumPy version < 1.19.

**Comment by sturlamolden** [comment]
Also it do not just affect Cobyla, but any function makes callbacks from Fortran to Python. A fix should perhaps be more global to SciPy and not just for Cobyla.

**Comment by pv** [comment]
Thanks, merged. We can consider rolling this back at some point.

The complication with Numpy version checks here is that we'd need to check the version of Numpy used to compile Scipy, not the currently imported one. Not sure if that information is stored somewhere. Because of the ABI compatibility stuff, the former is in practice going to often be quite old Numpy, so the f2py fixes won't come in soon. We perhaps should consider for the wheel builds upgrading f2py separately.

**Comment by sturlamolden** [comment]
f2py actually tries to report the f2py version used to compile the Fortran code in the docstring of the module. However, as of now this mechanism does not work. The import always fails, and the resported f2py version is always "2". If we just fix this tiny defect in f2py we can easily do this correctly.

https://github.com/numpy/numpy/blob/581eab37b5d558ba9b8035f3fa82cfa617e70e26/numpy/f2py/__version__.py

**Comment by tylerjereddy** [comment]
Does this warrant a backport label to prevent a segfault?

**Comment by sturlamolden** [comment]
Yes.

**Comment by sturlamolden** [comment]
It does not affect performance, Cobyla is currently single-threaded as the GIL is not released. The PR just protects the f2py wrapper code with an additional mutex, as it is not threadsafe (that is, unless we use f2py from NumPy master, but even if we did, the GIL would still serialize access to Cobyla).

**Pull request merged**